[DOCFIX] Add the memory overhead of rm/unmount/sync in the doc (#15448)
Alluxio/alluxio: docs/en/operation/User-CLI.md

* Adding `-R` option deletes all contents of the directory and the directory itself.

> * Adding `-U` option skips the check for whether the UFS contents being deleted are in-sync with Alluxio

Reviewer: sounds like we should make this the default option when we have a backcompat breaking release (ex. 3.0). i heard that @apc999  has a list of backcompat debt somewhere

Author: We are going to totally rewrite this part of logic (but this is not immediately happening).

$ ./bin/alluxio fs rm --alluxioOnly /tmp/unused-file2
```

When deleting only from Alluxio but leaving the files in UFS, you can use `-U` and `-Dalluxio.user.file.metadata.sync.interval=-1`
to skip the metadata sync and the UFS check. This will save time and memory consumption on the Alluxio master.
```console

> $ bin/alluxio fs rm -R -U --alluxioOnly -Dalluxio.user.file.metadata.sync.interval=-1 /path

Reviewer: ```suggestion
$ bin/alluxio fs rm -R -U --alluxioOnly -Dalluxio.user.file.metadata.sync.interval=-1 /dir
```
since you use `/dir` in the following examples (or change all of them to `/path`)

```

When deleting a large directory (with millions of files) recursively both from Alluxio and UFS,
the operation is expensive. 

We recommend doing the deletion in the following way:

> 1. Using the corresponding tool to check the UFS path directly and make sure everything can be deleted safely. 

Author: By "check" i mean that sanity check you would do before deleting a large dir in your own PC. It's probably not possible(or necessary) to check every file or subdir, but at least something can be done easily like looking at the top level, take a look, if anything looks interesting, step into that dir and take deeper look, and iterate until you feel confident.

At this part of doc I'd like to mention that `-U` does not really save your life, despite the fact that it's called a "check". All this `-U` does is to skip the files where the in-Alluxio and in-UFS views are different. This does not identify files you don't want to remove, nobody else can do that.

Also a metadata sync will not make the situation any better. Because knowing Alluxio is in sync with the UFS does not make anything better. You will still have to do the sanity check.

All I'm trying to add are mainly:
1. The UFS check (by not `-U`) is not your friend, neither is metadata sync.
2. You should really do the sanity check yourself.

For example if the UFS is HDFS, use `hdfs dfs -ls -R /dir` to list the UFS files and check.
Please do not sync the directory from Alluxio and check with `alluxio fs ls -R /dir`, because the loaded file metadata will
be deleted anyway and the expensive metadata sync operation will essentially be wasted.

> 2. Then issue the deletion from Alluxio and delete both in-Alluxio and in-UFS files:

Reviewer: ```suggestion
2. Issue the deletion from Alluxio to delete files from both Alluxio and the UFS:
```

```console
# Disable the sync and skip the UFS check, to reduce memory consumption on the master side
$ bin/alluxio fs rm -R -U -Dalluxio.user.file.metadata.sync.interval=-1 /dir
```

> If the metadata sync and UFS check are both disabled, deleting 1 million files from Alluxio will increase the JVM heap

Reviewer: this sounds too much like a story. let's just lay down the facts:
using an example of deleting 1 million files:
- if both metadata sync and ufs check are disabled, the jvm heap will increase by 2.5GB
- if only ufs check is disabled (but metadata syns is enabled?), the jvm heap will increases by 2.7GB (from 2.5GB +10%?)
- if both metadata sync and ufs check are enabled, the jvm heap will increase by 4.5 GB (2.5+2? i'm guessing because it says an extra 2 and 2 < 2.5, so i'm assuming i should add them together)

Author: When I was writing this part, my intention is to only recommend these 2 use cases:
1. Disable sync and check when `--alluxioOnly`
2. Disable sync and check when deleting UFS too

I tried to avoid mentioning too much about the other options because users should not use those and better not ask. And we are not too interested in giving a precise estimation on use cases we don't recommend, nor supporting(updating) those numbers in future versions.

That said, do you prefer my current description in the PR or listing the numbers as your recommendation? Both flavor look fine to me, and your math is correct.

Reviewer: okay great, then list those exact cases and the memory overhead for each with 1 million files as the example.

currently if i had to do the math for the second case, i have to take the proportion from the first example and then apply the 10%. and then when am i supposed to apply the 20% reduction mentioned for 2.8; was this accounted for already? now i am a terribly confused reader

Author: Ok i updated the section as recommended, see the new numbers, those are all for the latest 2.8 benchmark.

by around 2.5GB. If the UFS check is disabled by `-U`, deleting the files from UFS as well introduces around 10% extra overhead.

> Alluxio 2.8 introduced improvements to memory usage and reduced the JVM heap increase by around 20%.

Reviewer: the before and after version comparison belongs in the release notes, but not in documentation. realistically, the docs should never mention "starting from version x" or "y was added in version x".

Author: I'd like this make this as obvious as possible because I don't think people read the release note that attentively. Is there a better way?

Reviewer: but what are the chances that someone is reading this section as part of upgrading from pre-2.8 to post-2.8 or investigating the difference between as opposed to reading the release notes? there's never a good place to put it in the main docs; the release notes and other release related activities are the best places to point attention to version diffs. if people aren't reading the release notes and relying on searching through the individual doc pages, then we're not doing a good enough job of providing the version specific information.

another angle is what would the docs would look like if we listed all the "in version x, this changed" within the docs?

Author: Ok on 2nd thought I will remove this line from here, it belongs better to the release notes.

A digression - I think the user tends to just go to the latest doc page instead of going to the specific version first, so it might actually make sense in the latest version page to leave a pointer somehow.

Reviewer: no reader should need to jump back and forth to do the math. list the relevant numbers in both places.

Author: Ok removed the version and only 2.8 stats are kept

Metadata sync and UFS check will each add around 2GB extra overhead per 1 million files.

> So before issuing the recursive deletion command, measure the scale and estimate how much extra
> memory pressure this operation will add to the master.

Reviewer: ```suggestion
Using this example as a guideline, estimate the total additional memory overhead as a proportion to the number of files to be deleted. Ensure that the leading master has sufficient available heap memory to perform the operation before issuing a large recursive delete command.
```
